File: android/app/src/main/java/com/HammersTech/RoutineChart/core/domain/models/Routine.kt
package com.HammersTech.RoutineChart.core.domain.models

import java.time.Instant

/**
 * Routine domain model
 * Represents a sequence of steps to be completed
 * Routines are owned by a user (userId required) and optionally shared with a family (familyId optional)
 */
data class Routine(
    val id: String,
    val userId: String, // Owner of the routine (required)
    val familyId: String?, // Optional - if null, routine is personal; if present, routine is shared with family
    val title: String,
    val iconName: String?,
    val version: Int,
    val completionRule: CompletionRule,
    val createdAt: Instant,
    val updatedAt: Instant,
    val deletedAt: Instant?,
) {
    companion object {
        const val DEFAULT_VERSION = 1
    }

    val isDeleted: Boolean
        get() = deletedAt != null
}

File: ios/RoutineChart/RoutineChart/Core/Domain/Models/Routine.swift
//
//  Routine.swift
//  RoutineChart
//
//  Domain model for Routine entity
//

import Foundation

struct Routine: Identifiable, Codable, Equatable, Hashable, Sendable {
    let id: String
    let userId: String // Owner of the routine (required)
    var familyId: String? // Optional - if nil, routine is personal; if present, routine is shared with family
    var title: String
    var iconName: String?
    var version: Int
    var completionRule: CompletionRule
    let createdAt: Date
    var updatedAt: Date
    var deletedAt: Date?
    
    var isDeleted: Bool {
        deletedAt != nil
    }
    
    init(
        id: String = UUID().uuidString,
        userId: String,
        familyId: String? = nil,
        title: String,
        iconName: String? = nil,
        version: Int = 1,
        completionRule: CompletionRule = .all_steps_required,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        deletedAt: Date? = nil
    ) {
        self.id = id
        self.userId = userId
        self.familyId = familyId
        self.title = title
        self.iconName = iconName
        self.version = version
        self.completionRule = completionRule
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.deletedAt = deletedAt
    }
}

